addons/product/models/product.py

```diff
@@ -0,0 +1,879 @@
+                    result.append(_name_get(mydict))
+            else:
+                mydict = {'id': product.id,
+                          'name': name,
+                          'default_code': product.default_code}
+                result.append(_name_get(mydict))
+        return result
+
+    @api.model
+    def name_search(self, name='', args=None, operator='ilike', limit=100):
+        context = self.env.context
+        args = args or []
+        if name:
+            positive_operators = ['=', 'ilike', '=ilike', 'like', '=like']
+            ids = []
+            if operator in positive_operators:
+                ids = self.search([('default_code', '=', name)] + args, limit=limit).ids
+                if not ids:
+                    ids = self.search([('barcode', '=', name)] + args, limit=limit).ids
+            if not ids and operator not in expression.NEGATIVE_TERM_OPERATORS:
+                # Do not merge the 2 next lines into one single search, SQL search performance would be abysmal
+                # on a database with thousands of matching products, due to the huge merge+unique needed for the
+                # OR operator (and given the fact that the 'name' lookup results come from the ir.translation table
+                # Performing a quick memory merge of ids in Python will give much better performance
+                ids = self.search(args + [('default_code', operator, name)], limit=limit).ids
+                if not limit or len(ids) < limit:
+                    # we may underrun the limit because of dupes in the results, that's fine
+                    limit2 = (limit - len(ids)) if limit else False
+                    ids += self.search(args + [('name', operator, name), ('id', 'not in', ids)], limit=limit2).ids
+            elif not ids and operator in expression.NEGATIVE_TERM_OPERATORS:
+                ids = self.search(args + ['&', ('default_code', operator, name), ('name', operator, name)], limit=limit).ids
+            if not ids and operator in positive_operators:
+                ptrn = re.compile('(\[(.*?)\])')
+                res = ptrn.search(name)
+                if res:
+                    ids = self.search([('default_code', '=', res.group(2))] + args, limit=limit).ids
+            # still no results, partner in context: search on supplier info as last hope to find something
+            if not ids and context.get('partner_id'):
+                suppliers = self.env['product.supplierinfo'].search(
+                    [('name', '=', context.get('partner_id')), '|',
+                        ('product_code', operator, name),
+                        ('product_name', operator, name)])
+                if suppliers:
+                    ids = self.search([('product_tmpl_id.seller_ids', 'in', suppliers)], limit=limit).ids
+        else:
+            ids = self.search(args, limit=limit).ids
+        return self.browse(ids).name_get()
```

Comment:
use self.search() instead of self.search().ids because at the end you are re browsing ids